src/Sulu/Bundle/CategoryBundle/Controller/KeyWordController.php

```diff
@@ -0,0 +1,316 @@
+
+        $listBuilder->where($fieldDescriptor['locale'], $request->get('locale'));
+        $listBuilder->where(
+            $fieldDescriptor['categoryTranslationIds'],
+            $category->findTranslationByLocale($request->get('locale'))
+        );
+
+        // should eliminate duplicates
+        $listBuilder->distinct(true);
+        $listBuilder->addGroupBy($fieldDescriptor['id']);
+
+        $listResponse = $listBuilder->execute();
+
+        $list = new ListRepresentation(
+            $listResponse,
+            self::$entityKey,
+            'cget_category_key-words',
+            array_merge(['categoryId' => $categoryId], $request->query->all()),
+            $listBuilder->getCurrentPage(),
+            $listBuilder->getLimit(),
+            $listBuilder->count()
+        );
+
+        return $this->handleView($this->view($list, 200));
+    }
+
+    /**
+     * Creates new key-word for given category.
+     *
+     * @param int $categoryId
+     * @param Request $request
+     *
+     * @return Response
+     */
+    public function postAction($categoryId, Request $request)
+    {
+        /** @var KeyWord $keyWord */
+        $keyWord = $this->getKeyWordRepository()->createNew();
+        $category = $this->getCategoryManager()->findById($categoryId);
+        $keyWord->setKeyWord($request->get('keyWord'));
+        $keyWord->setLocale($request->get('locale'));
+
+        $keyWord = $this->getKeyWordManager()->save($keyWord, $category);
+
+        $this->getEntityManager()->persist($keyWord);
+        $this->getEntityManager()->flush();
+
+        return $this->handleView($this->view($keyWord));
+    }
+
+    /**
+     * Updates given key-word for given category.
+     *
+     * @param int $categoryId
+     * @param int $keyWordId
+     * @param Request $request
+     *
+     * @return Response
+     *
+     * @throws KeyWordIsMultipleReferencedException
+     * @throws KeyWordNotUniqueException
+     */
+    public function putAction($categoryId, $keyWordId, Request $request)
+    {
+        $keyWord = $this->getKeyWordRepository()->findById($keyWordId);
+
+        if (!$keyWord) {
+            return $this->handleView($this->view(null, 404));
+        }
+
+        $force = $request->get('force');
+
+        // overwrite existing keyword if force is present
+        if (null === $force
+            && $keyWord->isReferencedMultiple()
+            && in_array($force, [self::FORCE_OVERWRITE, self::FORCE_DETACH, self::FORCE_MERGE, null])) {
+            // return conflict if key-word is used by other categories
+            throw new KeyWordIsMultipleReferencedException($keyWord);
+        }
+
+        $category = $this->getCategoryManager()->findById($categoryId);
+
+        if ($force !== self::FORCE_MERGE
+            && $this->getKeyWordRepository()->findByKeyWord($request->get('keyWord'), $keyWord->getLocale()) !== null
+        ) {
+            throw new KeyWordNotUniqueException($keyWord);
```

Comment:
Wouldn't it be a better idea to put the throwing of these exceptions together with most of the logic here into the `KeywordManager`?